blackbird_python/blackbird/program.py

```diff
@@ -158,6 +162,66 @@ def operations(self):
         """
         return self._operations
 
+    @property
+    def parameters(self):
+        """List of free parameters the Blackbird script depends on.
+
+        Returns:
+            List[str]: list of free parameter names
+        """
+        return set([str(i) for i in self._parameters])
+
+    def is_template(self):
+        """Returns ``True`` if there is at least one free parameter.
+
+        Returns:
+            bool: True if a template
+        """
+        return bool(self.parameters)
+
+    def __call__(self, **kwargs):
+        """Create a new Blackbird program, with all free parameters
+        initialized to their passed values.
+
+        Returns:
+            Program:
+        """
+        if not self.parameters:
+            raise ValueError("Program is not a template!")
+
+        prog = copy.copy(self)
+        prog._parameters = [] # pylint: disable=protected-access
+
+        for op in prog._operations: # pylint: disable=protected-access
+            if 'args' not in op:
+                continue
+
+            for idx, a in enumerate(op['args']):
+                if isinstance(a, sym.Expr):
+                    par = list(a.free_symbols)
+                    func = sym.lambdify(par, a)
+
+                    try:
+                        vals = {str(p): kwargs[str(p)] for p in par}
+                    except KeyError:
+                        raise ValueError("Value for free parameter not provided")
+
+                    op['args'][idx] = func(**vals)
+
+            for k, v in op['kwargs'].items():
```

Comment:
same question here about these things being allowed to appear as both args and kwargs